Library/Homebrew/test/utils_spec.rb

```diff
@@ -0,0 +1,340 @@
+      end
+    end
+
+    describe "#which_all" do
+      let(:cmd1) { dir/"foo" }
+      let(:cmd2) { dir/"bar/foo" }
+      let(:cmd3) { dir/"bar/baz/foo" }
+
+      before(:each) do
+        (dir/"bar/baz").mkpath
+
+        FileUtils.touch cmd2
+
+        [cmd1, cmd3].each do |cmd|
+          FileUtils.touch cmd
+          cmd.chmod 0744
+        end
+      end
+
+      after(:each) do
+        dir.rmtree
+      end
+
+      it "returns an array of all executables that are found" do
+        path = [
+          "#{dir}/bar/baz",
+          "#{dir}/baz:#{dir}",
+          "~baduserpath",
+        ].join(File::PATH_SEPARATOR)
+        expect(which_all("foo", path)).to eq([cmd3, cmd1])
+      end
+    end
+
+    specify "#which_editor" do
+      ENV["HOMEBREW_EDITOR"] = "vemate"
+      expect(which_editor).to eq("vemate")
+    end
+
+    specify "#gzip" do
+      Dir.mktmpdir do |path|
+        path = Pathname.new(path)
+        somefile = path/"somefile"
+        FileUtils.touch somefile
+        expect(gzip(somefile)[0].to_s).to eq("#{somefile}.gz")
+        expect(Pathname.new("#{somefile}.gz")).to exist
+      end
+    end
+
+    specify "#capture_stderr" do
+      expect(
+        capture_stderr do
+          $stderr.print "test"
+        end,
+      ).to eq("test")
+    end
+
+    describe Shell do
+      subject { described_class }
+
+      describe "#shell_profile" do
+        before(:each) do
+          ENV["SHELL"] = shell
+        end
+
+        context "default" do
+          let(:shell) { "/bin/another_shell" }
+          its(:shell_profile) { is_expected.to eq("~/.bash_profile") }
+        end
+
+        context "Sh" do
+          let(:shell) { "/bin/sh" }
+          its(:shell_profile) { is_expected.to eq("~/.bash_profile") }
+        end
+
+        context "Bash" do
+          let(:shell) { "/bin/bash" }
+          its(:shell_profile) { is_expected.to eq("~/.bash_profile") }
+        end
+
+        context "Zsh" do
+          let(:shell) { "/bin/zsh" }
+          its(:shell_profile) { is_expected.to eq("~/.zshrc") }
+        end
+
+        context "Ksh" do
+          let(:shell) { "/bin/ksh" }
+          its(:shell_profile) { is_expected.to eq("~/.kshrc") }
+        end
+      end
+    end
+
+    describe "#popen_read" do
+      specify "without block" do
+        expect(Utils.popen_read("sh", "-c", "echo success").chomp).to eq("success")
+        expect($?).to be_a_success
+      end
+
+      specify "with block" do
+        expect(
+          Utils.popen_read("sh", "-c", "echo success") do |pipe|
+            pipe.read.chomp
+          end,
+        ).to eq("success")
+        expect($?).to be_a_success
+      end
+    end
+
+    describe "#popen_write" do
+      specify "with block" do
+        Utils.popen_write("grep", "-q", "success") do |pipe|
+          pipe.write("success\n")
+        end
+        expect($?).to be_a_success
+      end
+    end
+
+    specify "#pretty_duration" do
```

Comment:
Split these tests up if you can think of decent example names, but don't sweat it if not.